### ISSUE TITLE
Show window titles

### ISSUE THREAD

**Issue by xiangpeng2008** [kovidgoyal/kitty#4110]
Hi Kovid,

I'm just coming to kitty from tmux, because tmux doesn't currently support image printing while kitty could https://sw.kovidgoyal.net/kitty/graphics-protocol/.

Very often, I have multiple similar windows under same tab, which serve for similar but not exactly same jobs. And it's important for me to know which is which. With tmux (https://stackoverflow.com/a/37602055) I usually set each of the window a unique name. And show them on the bottom like below (I named them as "python 1" and "python 2"):
![Screenshot 2021-10-10 at 9 09 57 PM](https://user-images.githubusercontent.com/15676035/136697204-b8d83591-23fb-4621-975e-ae2d84e8c4f5.jpg)

But I didn't find this option in Kitty, do you think it worths to have this ?

Thanks,

Xiangpeng

**Comment by kovidgoyal** [comment]
No, I dont think so. In tmux it makes sense to do this as in tmux the
dividers use a cell height worth of pixels. In kitty they are only a few
pixels high so you cant have titles in them. You can however set window
titles, and the currently focused window's title will be show in the tab
bar.

**Comment by kovidgoyal** [comment]
On Sun, Oct 10, 2021 at 07:14:59AM -0700, xiangpeng2008 wrote:
> Thanks for your quick reply.

Doesn't seem worth the effort to me, sorry. Note that you can already
switch to the nth window by simply pressing ctrl+shift+n where n is 1-9
numbering from top left corner. Having to press a key to display the
numbers and then another to pick them seems much less efficient to me.

**Comment by xiangpeng2008** [comment]
How would you know the number of a specific window ?

**Comment by kovidgoyal** [comment]
On Sun, Oct 10, 2021 at 07:45:03AM -0700, xiangpeng2008 wrote:
> How would you know the number of a specific window ? 

they are numbered logically, just try it and see and you will get the
hang of it.

**Comment by xiangpeng2008** [comment]
I fully understand that it may not worth the effect to you as you may never need this.

On tmux press a key to display the numbers, then press the number "immediately" would focus the window, so it'
s still OK in terms of efficiency.

I find that the numbers could be a bit mixed if I do some 

Move window forward | ctrl+shift+f
-- | --
Move window backward | ctrl+shift+b

Which I do have to do so some times as I need to put the major working window in the biggest pane, and the major working window could change.

Thanks again for your help. Maybe I need to work around "Control kitty from scripts" to find another approach of my practice.

**Comment by kovidgoyal** [comment]
On Sun, Oct 10, 2021 at 08:05:58AM -0700, xiangpeng2008 wrote:
> I fully understand that it may not worth the effect to you as you may never need this.

Indeed, but a PR implementing it is most welcome.

> 
> On tmux press a key to display the numbers, then press the number "immediately" would focus the window, so it'
> s still OK in terms of efficiency.

Of course, I am not saying it doesnt work for you, just doesnt make
sense for me.

> 
> I find that the numbers could be a bit mixed if I do some 
> 
> Move window forward | ctrl+shift+f
> -- | --
> Move window backward | ctrl+shift+b
> 
> Which I do have to do so some times as I need to put the major working window in the biggest pane, and the major working window could change.
> 
> Thanks again for your help. Maybe I need to work around "Control kitty from scripts" to find another approach of my practice.

If you are willing to do a bit of scripting, you can do the following:

map f1 launch --type=overlay --allow-remote-control my-window-switcher-script

And in the switcher script you use

kitty @ ls to get a list of window titles and have the script ask the
user to press a number key to select the window to switch to.

**Comment by kovidgoyal** [comment]
that's because the released version of the installer has not been
updated for nightly builds yet, use

https://github.com/kovidgoyal/kitty/raw/master/docs/installer.sh

as the installer url instead

**Comment by kovidgoyal** [comment]
On Tue, Oct 12, 2021 at 08:23:47AM -0700, xiangpeng2008 wrote:
> great, this url works to install a new kitty
> what should I do to show the window numbers and press key to focus ?
> I'm too bad, I looked at your commit but I don't know how let kitty run "focus_visible_window" ...

ctrl+shift+f7

> 
> Actually I have a general question:
> this is the mapping for previous_window
> `map ctrl+shift+[ previous_window`
> Is there anyway I just type "previous_window" to go to previous window ? In case I don't know the keyboard shortcut.

no

**Comment by kovidgoyal** [comment]
You need to have more than one visible window for it to work. And read the changelog for all the other changes. https://github.com/kovidgoyal/kitty/issues/3948

**Comment by page-down** [comment]
Sorry, is this a little off topic? Since it's already been discussed here, I'll make two more suggestion.

If the focus is switched using a mouse click while the numbers are displayed, can we assume that the user has made a selection and hide the number display? Although no keyboard keys are pressed.

It would be nice to be able to use this function to also learn about the windows in the layout of the `stack` type. For example, a quick preview of the windows (maybe titles only) present in the stack and displaying the numbers.

I can understand the need for the title proposed in this post. It would be convenient to list the titles of the windows in stack mode and switch between them with numbers.

**Comment by kovidgoyal** [comment]
The point of this feature is, you press the number for the position you want to go to. It doesnt matter what the number is. If you want to always go to some specific window use its id or title.

**Comment by kovidgoyal** [comment]
See https://github.com/kovidgoyal/kitty/commit/533b0ed591b75103a26b811aab01831d31e51d55

**Comment by kovidgoyal** [comment]
There isn't any way to do that, if you want to do that you will need to
implement a proper kitten to do it instead of using the hints kitten as
is done now.

**Comment by kovidgoyal** [comment]
See https://sw.kovidgoyal.net/kitty/layouts/#the-splits-layout

**Comment by xiangpeng2008** [comment]
> On Tue, Oct 12, 2021 at 08:23:47AM -0700, xiangpeng2008 wrote: great, this url works to install a new kitty what should I do to 
> Actually I have a general question: this is the mapping for previous_window `map ctrl+shift+[ previous_window` Is there anyway I just type "previous_window" to go to previous window ? In case I don't know the keyboard shortcut.
> no

Do you think it make sense to enable kitty remote control to do jobs which could only been currently done with Shortcut ?
For example when I'm focusing window num 2, I could do send a system command like 
`kitty @ action --match num: 3 move_window_to_top`
which would do the same as 
1.  I focus num 3
2.  I press ctrl+shift+`
3.  I focus back to num 2

This could enable more automation and customizing on kitty control which could only been done previously with Keyboard Shortcut.
For example, if I need to do above action a lot, then I could just map a new shortcut like
`map ctrl+p>3  kitty @ action --match num: 3 move_window_to_top`

**Comment by xiangpeng2008** [comment]
> See https://sw.kovidgoyal.net/kitty/layouts/#the-splits-layout

Is there a specific reason 
only move_window + [ left, right, up, down ]  works for "splits", while move_window_forward or backward doesn't ?

It seems we are separating "splits" and other layouts here: 
https://github.com/kovidgoyal/kitty/blob/master/kitty/tabs.py#L562-L567

But seeing any layouts except "stack" could be constructed by "splits" would it make sense to make "splits" as super class of all other layouts except "stack" ?

Also currently "move_window_to_top" doesn't work with "splits" layouts. Besides could we make swap_windows https://github.com/kovidgoyal/kitty/blob/master/kitty/layout/splits.py#L70 operable with keyboard shortcuts ? Like `swap_window 1 2` would swap windows 1 and 2.

**Comment by kovidgoyal** [comment]
Because there is no forward and backward in a split. Nor is there a top.

**Comment by kovidgoyal** [comment]
It is tree like, but I don't see navigating the branches of a tree in
that fashion as being meaningful. The problem is that windows are visual
things, and forward/backward/top make some sense for linear visual
structures (which the other layouts are) but not really for tree like visual structures.

You typically move focus to get to some place. In the splits it is not
at all obvious what sequence of movements will get you to a window,
unless you have the tree structure in front of you, which you dont in
kitty, since a given visual window layout can correspond to many
different underlying tree structures.

**Comment by xiangpeng2008** [comment]
> You could, but what would be the point? Unless you memorise the order of creation, it will just jump to random places.

Sorry, order of creations may not be good.

But what @page-down  said makes sense to me. It would be quite natural for a user to know which window comes first.

For example :

![image](https://user-images.githubusercontent.com/15676035/137583525-7ba97f4f-fec9-44c6-a91b-7243348efcaf.png)

**Comment by page-down** [comment]
> it's very natural to know that it would go
Not really, because any window can be split in two in kitty. And those hidden virtual groups are not visible and have no borders.

Because the current structure of splits is so free, it doesn't make sense to move nodes around. And it's a visual problem.

Unless a very simple structure is introduced, that is, when doing a split, the window is actually added to the parent virtual group.
A new virtual group is created only when the direction of the split is different from the parent.

```text
- group[v]
  - window-1
  - group[h]
    - window-2
    - window-3 <-- hsplit

- group[v]
  - window-1
  - group[h]
    - window-2
    - window-3
    - window-4 <-- (new)
```

When creating windows with different directions, create a new virtual group and set the split direction.
```text
- group[v]
  - window-1
  - group[h]
    - window-2
    - window-3 <-- vsplit (does not match parent group "horizontal")

- group[v]
  - window-1
  - group[h]
    - window-2
    - group[v] <-- (new)
      - window-3
      - window-4 <-- (new)
```

Yet this still does not solve the visual problem.

For example, if you have a layout of 4 windows, how do you know how they will move when you execute move forward? The split direction and virtual group boundaries are not shown on the screen.

```text
- group[h]
  - group[v]
    - window-1
    - window-2
  - group[v]
    - window-3
    - window-4
```

So to solve this problem is not something that can be done by simply moving. Because the result after pressing the shortcut is probably not what the user expects. The visual interface problem has to be solved first, otherwise it is meaningless to move blindly.

**Comment by xiangpeng2008** [comment]
> 
> For example, if you have a layout of 4 windows, how do you know how they will move when you execute move forward? The split direction and virtual group boundaries are not shown on the screen.
> 
> ```
> - group[h]
>   - group[v]
>     - window-1
>     - window-2
>   - group[v]
>     - window-3
>     - window-4
> ```
> 
> So to solve this problem is not something that can be done by simply moving. Because the result after pressing the shortcut is probably not what the user expects. The visual interface problem has to be solved first, otherwise it is meaningless to move blindly.

This is a very special case, when there's a + shape cross on border. It's true that it's hard to tell between 1 and 2. But the point here it's not random for the user, and if user used to split windows in same way, like always h->2v or v->2h, then he knows the order of windows.
![image](https://user-images.githubusercontent.com/15676035/137584500-c5399ea5-2048-4c16-b6e8-0430920d03c9.png)
Besides above case could also appear like this, whose order is very clear
![image](https://user-images.githubusercontent.com/15676035/137584574-96aff8f4-4d64-4df5-a41a-404369f12042.png)

Btw instead of simply moving, switch could be more convenient on some user case, for example, when I have 3 monitors, and I want to use window 0 on left screen to view output, window 1 on middle screen for vim editing, and some opened window on right window. And when I want to read outputs of window 5, it would be convenient to just switch between window 0 and window 5.
![image](https://user-images.githubusercontent.com/15676035/137584682-1963bf84-431e-4e19-86ed-035a46302c6d.png)

**Comment by page-down** [comment]
Wow! This is a piece of art. Many thanks for your work. For the better, I have some suggestions for a different point of view and I hope you don't mind.

I really like the seven-segment numeric display. However, there are some balance and word weight issues from a typographic perspective. For example, the number `1` is not centered. (And it is also not appropriate to force centering in a seven-segment numeric display, as it is not logical.) Also the number `4` looks smaller than the other numbers because of the lack of top and bottom LEDs. In a world of monospaced fonts, is there a better way to display them? I don't know if there is a better way to display it with the least amount of effort.

When trying different layouts, I sometimes want to limit the layout used by the tab and the layout cycled by `ctrl+shift+l`. It would be more convenient to have a remote control command to set the layout that can be used on-the-fly.

We already have `KeyAction('set_tab_title')`, would you consider adding a `KeyAction('set_window_title')` to complement this functionality? Thanks.

**Comment by kovidgoyal** [comment]
On Sun, Oct 17, 2021 at 02:44:41AM -0700, page-down wrote:
> Wow! This is a piece of art. Many thanks for your work. For the better, I have some suggestions for a different point of view and I hope you don't mind.

You are welcome!

> 
> I really like the seven-segment numeric display. However, there are some balance and word weight issues from a typographic perspective. For example, the number `1` is not centered. (And it is also not appropriate to force centering in a seven-segment numeric display, as it is not logical.) Also the number `4` looks smaller than the other numbers because of the lack of top and bottom LEDs. In a world of monospaced fonts, is there a better way to display them? I don't know if there is a better way to display it with the least amount of effort.
> 

You could obviously render the musing a font rather than a seven segment
display, but I like the seven segment display. Both of those issues are
"natural" in a seven segment display.


> When trying different layouts, I sometimes want to limit the layout used by the tab and the layout cycled by `ctrl+shift+l`. It would be more convenient to have a remote control command to set the layout that can be used on-the-fly.

Sure, this is easy enough, but open a separate issue to track it.

> 
> We already have `KeyAction('set_tab_title')`, would you consider adding a `KeyAction('set_window_title')` to complement this functionality? Thanks.

I have no objection, though I also don't really have a need for it, since
there are standard escape codes to set window titles. The only possible
use case is if you want to set titles of another window than the one you
are running in. Which is not something I have ever needed to do.

**Comment by xiangpeng2008** [comment]
wow, we have title now

<img width="613" alt="Screenshot 2021-10-30 at 2 27 36 PM" src="https://user-images.githubusercontent.com/15676035/139523026-6e83e48e-e664-4e1f-aac0-fdf5e48cc796.png">

**Comment by elzii** [comment]
I stumbled across this issue -- I made a solution akin to tmux's `prefix + s` awhile back: 

**Dependencies:**
- `bash`
- `fx` (`npm install -g fx` -- similar to `jq`) 
- `fzf`
- `column` from `util-linux` (`/opt/homebrew/opt/util-linux/bin/column`)

**Assumptions:**
- Your login shell is `zsh` from `homebrew` (can change this part easily -- most the logic is in parsing `kitty @ ls`
- You launch `Kitty.app` with `--single-instance` and `--allow-remote-control`


**Screenshot**:

<img width="665" alt="ZIZZOCAPTURE--2024-10-13@06 08 33-PM" src="https://github.com/user-attachments/assets/cf575fb5-6f23-4159-9c86-cd9c67d16fde">


**Bash Script:**

```sh
#!/usr/bin/env bash

PROG=$(basename $0)

[[ -n $HOMEBREW_PREFIX  ]] && { BIN_DIR=$HOMEBREW_PREFIX/bin; } || { BIN_DIR=/opt/homebrew/bin; }

function procs() {
  $BIN_DIR/kitty @ ls \
    | $BIN_DIR/fx "x => x.map(w => w.tabs.map(t => t.windows)).flat().flat().map(t => t.foreground_processes.map(fg => ({ id: t.id, pid: fg.pid, cwd: fg.cwd.replaceAll(/\/Users\/azizzo/g, '~'), cmd: fg.cmdline[0].replaceAll(/-/g, '')}) )).flat()"
}

function tabs() {

  $BIN_DIR/kitty @ ls \
    | $BIN_DIR/fx 'x => x.map(n => n.tabs)
      .flat()
      .map(t => ({ 
        tabId: t.id, 
        focused: t.is_focused, 
        title: t.title.length > 16 ? t.title.substring(0, 16)+"..." : t.title, 
        winCmd: Object.values(t.windows[0].cmdline)[0].replace(/^-/g,""), 
        winCwd: Object.values(t.windows[0].cwd)[0], 
        winId: parseInt(Object.values(t.windows[0].id)[0], 10), 
        pid: Object.values(t.windows[0].pid)[0] 
      }))' \
    | sed -E 's/undefined//g' 

}


function choose() {
  tabs \
    | $BIN_DIR/jq -r '.[] | [.tabId, .title, .winCmd, .pid] | @tsv' \
    | /opt/homebrew/opt/util-linux/bin/column -t -s $'\t' -o $'\t' -N 'ID,FOCUSED,TITLE,CMD,PID' \
    | /Users/azizzo/.bin/colorize -d $'\t' | head -n -1 \
    | grep -v 'KITTYCHOOSE' \
    | $BIN_DIR/fzf --ansi -d $'\t' --layout=reverse --info=inline -i --multi --header-lines=1 \
      --header "<Return> = Focus | <ALT-Return> = Print selected line(s) | <Alt-c> Copy PID" \
      --bind "return:execute(kitty @select-window --match id:{1})+abort" \
      --bind "alt-return:execute(echo {5})+abort" \
      --bind "alt-c:execute(echo {1} | pbcopy)+abort"
}


function usage(){
  local title=$(printf "\033[30m\033[47m %s \033[49m\033[39m" $PROG)

  cat <<< "
  ${title}
  FLAG                  FLAG DESCRIPTION
  ▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔
  --help,     -h        Show this menu
  --procs               List kitty by process
  --tabs                List kitty tabs
  --choose              Choose kitty tab to navigate to
  "
  return 0
}


if [[ -n $@ ]]; then
  case $@ in
    *--tabs*)
      shift
      tabs
      ;;
    *--procs*)
      shift
      procs
      ;;
    *--choose*)
      shift
      choose
      ;;
    *--help*)
      shift
      usage
      ;;
    *)
      usage ;;
  esac
else
  usage
fi
```



**Final Steps**
Try `kitty-list --choose`, or for more convenience -- in your config, add:

```sh
# Change '~/.bin/kitty-list' to whereever you have the script in your $PATH
map cmd+alt+s    launch --allow-remote-control --type=overlay --title KITTYCHOOSE --tab-title KITTYCHOOSE --copy-env --cwd ~ ~/.bin/kitty-list --choose
```

Cheers. Love kitty, @kovidgoyal. IMO underrated and so OP compared to Alacritty, WezTerm, et al.

Wish we had good community portal for folks to share their kittens / scripts that complement it, I have so many just lying around.
